[DDW-665] Rewrite numeric input logic

DominikGuzei opened the pull request:
This PR rewrites and simplifies the logic of the `NumericInput` component.

Fixes #112 and fixes #51

## Breaking API changes:
  - `value` must be a number and `onChange` also returns numbers (not strings) now
  - All formatting configuration is now handled with the prop `numberLocaleOptions` which are the options for [Number::toLocaleString](https://developer.mozilla.org/de/docs/Web/JavaScript/Reference/Global_Objects/Number/toLocaleString), which makes this component more standards compliant and will also allow us to implement i18n later on without too much hassle.
- The numeric input component does **not handle errors or enforce min/max values anymore** - since error messages created by the numeric input would never be useful in an localized context (and the enforced versions, which dont allow out-of-bound numbers to be entered at all, are not very useful UX wise -> because it's better to see the "invalid" number + a user defined error)

## Specification

Since there is no standard on how to build numeric input components, here is the specification we came up with that serves our purposes in the best way: 

- Only numeric inputs that are representable by Javascript numbers are valid. This is guarded by `Number.MIN_SAFE_INTEGER` 
  (-9007199254740991) and `Number.MAX_SAFE_INTEGER` (9007199254740991) but since also fractions need to 
  represented, the calculation for the maximum integer part goes like this:
  `Number.MAX_SAFE_INTEGER / 10 ** (maximumFractionDigits + 1)` (which basically means that one integer digit is lost for
  each supported fraction digit). For `maximumFractionDigits == 3` this results in 
  `9007199254740991 / 10 ** 4 == 900719925474.099` being the biggest number that can be entered.
- Only numeric digits `[0-9]` and dots `.` can be entered.
- When a second dot is entered it replaces the existing one and updates the fraction part accordingly
- Commas cannot be deleted but the cursor should jump over them when DEL or BACKSPACE keys are used
- The fraction dot can only be deleted if `minimumFractionDigits` is not defined or
  if the resulting number does not exceed the numeric limits!
- It should be possible to replace the whole number or parts of it (even the dot) by inserting another number.
- If the fraction dot is deleted but the resulting number is too big the cursor jumps over the dot without deletion
- If you insert a digit but the resulting number would exceed the numeric limit, nothing happens


## How to Test

Discuss this change in the corresponding [daedalus-qa thread](https://input-output-rnd.slack.com/archives/GGKFXSKC6/p1560350401014900)

- Use the storybook build below (Github PR checks)
- Enter any numeric number and choose different scenarios in storybook

![screenshot 2019-06-12 um 16 43 12](https://user-images.githubusercontent.com/172414/59360919-44481780-8d31-11e9-9e59-41922d345e2f.png)

nikolaglumac commented:
@DominikGuzei I have added WIP tag to this PR until the issues are resolved.

DominikGuzei commented:
Thanks for the reviews and feedback @MarcusHurney  and @tomothespian - i will work on these issues over the weekend!

@MarcusHurney one thing i can tell you right away: the number of digits that can be entered will be limited to the Javascript number limits (but still it's not correct in this case since Javascript supports up to 21 digits)

DominikGuzei commented:
@MarcusHurney @tomothespian all mentioned issues have been fixed!

And yes … the 3 max fraction digits is the Javascript default 😉

DominikGuzei commented:
Great catch @tomothespian … to explain why this happens: The reason why the numbers increased for any last fraction digit bigger than 4 was that `Number::toLocaleString` rounds the number to the precision … but now you discovered another rounding edge-case: if there are no more bits to represent the number (in your case the integer part already ate up most of the 64 bit and thus no more digits could be supported and the rounding issue happens again 😓 will try to find a fix for that

DominikGuzei commented:
@tomothespian i have fixed the edge-case you found! It should not be possible to enter an invalid number anymore 👍

DominikGuzei commented:
@tomothespian that's expected since the numeric input (in the new form) is limited by the numeric limits in Javascript (64 bits floating point representation of numbers) … i think that's acceptable since in reality you rarely need to deal with such big integer numbers. The reason why i don't want to introduce arbitrary precision (e.g big.js) is that makes formatting the numbers etc. more complicated (and we probably won't need it anyway)

IuliaDolishniak commented:
Is this expected when I type over the existent number another needed number it doesn't change and stays the same, i.e. when there is 1,111,111,111,111,111 already typed and I type over another number, e.g. 2 it stays the same 1,111,111,111,111,111?

nikolaglumac commented:
@IuliaDolishniak check this:
https://github.com/input-output-hk/react-polymorph/pull/114#pullrequestreview-256780421
...and this comments:
https://github.com/input-output-hk/react-polymorph/pull/114#issuecomment-507626911

IuliaDolishniak commented:
I have read this and it is mainly for decimals that's why I wanted to double check as this is confusing.

nikolaglumac commented:
@IuliaDolishniak I think that doesn't matter but it would be good if @DominikGuzei can confirm this 🙏

DominikGuzei commented:
Great catch @IuliaDolishniak, i pushed a fix to this bug!

IuliaDolishniak commented:
I still can see this issue when I try to change the number by entering another one the cursor just moves but number doesn't change.
![input](https://user-images.githubusercontent.com/47106784/60537613-961d0580-9d11-11e9-86f7-e85ebd0c146d.gif)

nikolaglumac commented:
@DominikGuzei can you please check this issue @IuliaDolishniak reported? 🙏

DominikGuzei commented:
@IuliaDolishniak im not seeing the issue you mentioned on the latest version - please double check the latest version (or if you still see the issue please show me more exactly how to reproduce it)

DominikGuzei commented:
@IuliaDolishniak ah i just reproduced the issue you're describing, will work on it! 👍

DominikGuzei commented:
@nikolaglumac ok i found the reason … but it's not really a bug but more a UX behavior that we need to think about: the current implementation does not overwrite digits but inserts new digits where you place your cursor (which is a pretty standard UX behavior IMO) … but what that means is that if you have a really big number (near the limit) you cannot insert another digit because then you would get a number that is bigger than the limit … and thus it is not accepted in this case. If you want to replace a digit you have to select it and then you can replace it with any other number.

But i see the "bug" in the current implementation is that the cursor is still moved although nothing is entered … i will fix that now!

DominikGuzei commented:
@DmitriiGaico @nikolaglumac @tomothespian @IuliaDolishniak ready for final round of reviews!

gnpf commented:
@DominikGuzei 
I found the following:
Having inserted 1,000,000,000,000,000 _(or 2,000,000,000,000,000 // 3,000,000,000,000,000, etc)_ try to insert any number before or after any zero. Similar to what @IuliaDolishniak reported, the cursor changes position (to the left), even when no number can be inserted. 

![LFhGwj9ZKx](https://user-images.githubusercontent.com/14169859/60919399-fc52dc80-a26b-11e9-98b9-3ef89fd47652.gif)

DominikGuzei commented:
@MarcusHurney @gnpf @DmitriiGaico i fixed the issues you mentioned about not being able to insert certain fraction digits or the cursor being moved without changing digits!

The logic is this now: Depending on the configured (or default = 3) `maximumFractionDigits` the integer part of the number is limited in a way that you cannot run into number overflow issues (that was the problem with fraction parts not being accepted or jumping with the cursor without input).

The downside is that if we have `maximumFractionDigits = 6` then the largest number you can enter is `999,999,999.999999` (since Javascript cannot represent more information in one number)

DominikGuzei commented:
The other option we have is to dynamically adjust the integer limit depending on the entered fraction digits … so with every fraction digit that we need to capture the precision of the number you can enter one integer digit less … the question is which is better UX wise. I think the fixed solution (explained above) is easier to understand, since its not dynamic … but the downside is that you always reserve all bits of information for the fraction part, which might not be needed by the user.

The dynamic variant is more flexible but the logic behind might not be obvious to the end user.

gnpf commented:
@DominikGuzei The fix looks good.
Regarding the different options, I'm not sure if the user would need more precision or be able to insert bigger numbers, but I do agree that the dynamic option is less intuitive than the one that only allows 3 decimals, so I would choose the latter.

gnpf commented:
@DominikGuzei 
I found another issue: On the decimals, it is not possible to add zeros to the left unless you add another digit first. For example, if I want to add _123.106_ I have to insert _123.16_, then place the cursor between the _1_ and _6_ to insert the zero. 
Maybe it would be more friendly if the zeros can be inserted to the end of the decimal and deleted (if no numbers are after them) but just after the user change the focus or after a few seconds have passed.

DominikGuzei commented:
Great catch @gnpf  👏 i fixed the issue and improved the general handling of entering fractions - please take it for another deep test drive!

DominikGuzei commented:
Please don't merge yet, i found another issue when you enter `0.0` weird stuff is happening … need to investigate

nikolaglumac commented:
@DominikGuzei please let us know once the issues are fixed. 
Also, please share the issue details so that @DmitriiGaico and @gnpf can re-test it after the fix is in place. Thanks!

DmitriiGaico commented:
@DominikGuzei Just tried 100.00 combination. It also fails when you hit first 0 after the dot :(

DominikGuzei commented:
@DmitriiGaico @gnpf @nikolaglumac i fixed the issue and trailing zeros in the fraction part are now supported 👍

DmitriiGaico commented:
@doinikGuzei Hi Dominik. The 0.0 looks like fixed but i found a new issue.
Here is the bug scenario.
Enter 5,555,555,555,555.000
Now you cant enter any number because maximum is reached.
Erase all the 000
Place pointer between any of the fives and you can enter more numbers, you can enter even characters.
Thank you.

DominikGuzei commented:
@DmitriiGaico i cannot reproduce your last bug report! Maybe you're on a outdated / wrong netlify build?

DominikGuzei commented:
@DmitriiGaico ok sorry, now i could reproduce it 👍

DominikGuzei commented:
@DmitriiGaico i fixed the problem 👍

DominikGuzei commented:
@input-output-hk/daedalus-review ready for another round of reviews 🌮

nikolaglumac commented:
@DominikGuzei Dmitrii is on vacation this week. @IuliaDolishniak @gnpf can one of you please check the fix for the issue Dmitrii reported? 🙏

IuliaDolishniak commented:
@DominikGuzei found 2 issues 
1.  when I enter the number starting with any number from 1 to 8 I can enter a trillion, i.e. the number with 12 zeroes when I enter number starting with 9 I am only able to enter the number with 9 zeroes, i.e. billion. Not a big problem but there is a discrepancy that I should let you know of
![numbers](https://user-images.githubusercontent.com/47106784/61234976-16ddf780-a73d-11e9-952a-b42ffb9d20cc.gif)
2. when I enter number containing of only 8s I got the last number transformed into 9, also not big issue
![numbers_2](https://user-images.githubusercontent.com/47106784/61235069-44c33c00-a73d-11e9-8822-59654f6cc715.gif)

nikolaglumac commented:
@DominikGuzei please check the comment @IuliaDolishniak posted and be sure to add all the numeric limitations (related to the JS handling) to the docs 🙏

gnpf commented:
@DominikGuzei The fix for the issue @DmitriiGaico reported is working fine.
I found some not so friendly behaviors that maybe can be improved, let me know what you think,

* Placed after a comma, you can't delete the number before it (possible if placed before the comma).

* Placed after the dot (with maximum digits reached) you can't delete the number before it, the cursor just jumps a digit.

![e8dOnhIByc](https://user-images.githubusercontent.com/14169859/61242877-2baa9300-a71d-11e9-9a7c-6fc266b6cc04.gif)

* Placed before the dot you can't delete the number after it, the cursor just jumps a digit.

![VXxDqWfstm](https://user-images.githubusercontent.com/14169859/61243622-c0fa5700-a71e-11e9-8850-9a4294dac545.gif)

DominikGuzei commented:
@DmitriiGaico i fixed the issues you mentioned!
@IuliaDolishniak i added the specification of the component to the PR description, what you can see there is that it is expected that you can enter any number up to `9007199254740.99` with a maximum of three fraction digits configured!

IuliaDolishniak commented:
@DominikGuzei can you please double check the second issue? As there when I enter 8 after the dot is transformed into 9. This is not very obvious from the video but this is confusing as by me.
And the first one shows the difference between the number that starts with 9 and can be a billion, but if it start with 8 it can be a trillion. Please let me know If this is expected and we can live with it.

DominikGuzei commented:
@IuliaDolishniak sorry i forgot to push a last change that fixes the issue of the last digits swapping in some edge cases 👍

DominikGuzei commented:
Please do not merge yet 🙏 i found another issue when minimum fraction digits is set

DominikGuzei commented:
Ok false alarm, everything is fine - just forgot to rebuild the project locally 👍

DominikGuzei commented:
Ok while testing on Daedalus i found another issue when deleting the decimal point in the `minimumFractionDigits` scenario 😓

DominikGuzei commented:
@gnpf and @IuliaDolishniak ready for another testing 🙏

IuliaDolishniak commented:
@DominikGuzei 

> The fraction dot can always be deleted (which transforms the fraction part into integer digits)
except if the resulting number would exceed the numeric limits!

>  
I cannot delete the fraction dot, cursor just moves forward and nothing happens.

DominikGuzei commented:
@IuliaDolishniak good catch, i fixed the regression! But i also updated the specification to include that when `minimumFractionDigits` is defined it should not be possible to delete the dot (for obvious reasons). Ready for re-test 👍

DominikGuzei commented:
@MarcusHurney @SebastienGllmt please checkout this PR as this will replace the current `NumericInput` component in the next release 🙏 the biggest difference is that it now returns Number instead of String and that the configuration works differently.

DominikGuzei commented:
@IuliaDolishniak @gnpf i fixed another issue: when `minimumFractionDigits` was defined it was not possible to select the whole number or parts of it (if the dot was included) and replace it with another number. Now that should be possible 👍 The dot just cannot be removed by deletion

Please re-test 🙏

nikolaglumac commented:
@gnpf @DmitriiGaico can you please test and approve this one? 🙏

IuliaDolishniak commented:
@DominikGuzei as a user I would like when all decimals are removed then the dot also so should disappear. I don't think it is right to leave the number look like this `12345455.`
Can we have it removed itself when all the number after dot are removed?
![screen](https://user-images.githubusercontent.com/47106784/62777951-b3cf4e80-bab7-11e9-92e7-9c3ebd5aea26.gif)

gnpf commented:
@DominikGuzei I've been checking the last updates and have a few questions:

1 - The maximum number of decimals digits allowed is now 3. Is that part of the changes you implemented or is it a bug? Last time I checked it was 6
2 - I'm not sure if I understand the issue you mentioned. When you said "The dot just cannot be removed by deletion" is that the previous behavior or the expected after the fix?
3 - Should we wait for this to be merged to start with DDW-754?

SebastienGllmt commented:
We have one of our engineers testing this for Yoroi also. You can find the PR here: https://github.com/Emurgo/yoroi-frontend/pull/888

gnpf commented:
@DominikGuzei Also found that the maximum supported number is now smaller by 1 digit
It used to be  9.000,000,000,000 and now is 900,000,000,000.

nikolaglumac commented:
@gnpf doesn't that fall under the specification @DominikGuzei wrote?
![Screenshot 2019-08-09 at 15 55 59](https://user-images.githubusercontent.com/376611/62784205-3a872a00-babe-11e9-9f57-70609ada1ea6.png)
We have agreed that this specification is what we need. So please make sure to compare your test results with this specification. There are some differences in comparison to how the previous version worked.

gnpf commented:
> @gnpf doesn't that fall under the specification @DominikGuzei wrote?
> ![Screenshot 2019-08-09 at 15 55 59](https://user-images.githubusercontent.com/376611/62784205-3a872a00-babe-11e9-9f57-70609ada1ea6.png)
> We have agreed that this specification is what we need. So please make sure to compare your test results with this specification. There are some differences in comparison to how the previous version worked.

You're right, I missed that part. Ignore my previous comment.

gnpf commented:
@DominikGuzei Not a big deal, but when pressing backspace after the comma or delete before the comma, the first time the comma is jumped over, but after the first attempt on the same comma, the cursor will go to the end of the number;

![ZTdRHwtfVr](https://user-images.githubusercontent.com/14169859/62804817-bab08e00-bac4-11e9-9b6a-9c8684e65fe6.gif)

DominikGuzei commented:
@gnpf 
1 - The maximum number of decimals digits allowed is now 3. Is that part of the changes you implemented or is it a bug? **This is the expected default value, can be adjusted (like in Daedalus)**
2 - I'm not sure if I understand the issue you mentioned. When you said "The dot just cannot be removed by deletion" is that the previous behavior or the expected after the fix? **That was changed, the dot can be deleted but only if no `minimumFractionDigits` is defined.**
3 - Should we wait for this to be merged to start with DDW-754? **No, this PR has to be merged and released before DDW-754**

IuliaDolishniak commented:
> @DominikGuzei as a user I would like when all decimals are removed then the dot also so should disappear. I don't think it is right to leave the number look like this `12345455.`
> Can we have it removed itself when all the number after dot are removed?
> ![screen](https://user-images.githubusercontent.com/47106784/62777951-b3cf4e80-bab7-11e9-92e7-9c3ebd5aea26.gif)

@DominikGuzei can you pleas comment the issue above?

DominikGuzei commented:
@IuliaDolishniak great catch 🙏 i fixed the issue and you can re-test!

DominikGuzei commented:
@IuliaDolishniak i will try to fix the problem with dangling dot! 👍

DominikGuzei commented:
@IuliaDolishniak i fixed the issue with the dangling dot when all fractions have been deleted! 👍 ready for re-test

gnpf commented:
> @gnpf great catch 🙏 i fixed the issue and you can re-test!

Checked this and is working fine, great job 👍

nikolaglumac commented:
@MarcusHurney please review this one 🙏

nikolaglumac commented:
@DominikGuzei I have put a WIP label on this PR until the issues @MarcusHurney found are fixed: https://github.com/input-output-hk/react-polymorph/pull/114#pullrequestreview-274002383

DominikGuzei commented on source/components/NumericInput.js:
```diff
@@ -110,385 +99,342 @@ class NumericInputBase extends Component<Props, State> {
+
+// ========= HELPERS ==========
+
+const VALID_INPUT_REGEX = /^([0-9,\+\-\.]+)?$/;
```
This reminds me that we need to fix the husky scripts as it seems like they are not working correctly to prevent pushing invalid code 😉
